### ISSUE TITLE
raise warning/error or use WKT for WMTS boundingbox CRS when no URI exists

### ISSUE THREAD

**Issue by AndrewAnnex** [developmentseed/titiler#1043]
So an issue I've known would be coming my way is in the use of `CRS_to_uri` inside the WMTSCapabilities.xml endpoints for the bounds object CRS parameter, which when using a slightly tweaked/custom CRS can result in a `None` being specified in the WMTS XML document. This causes issues with clients like QGIS, but will work as soon as a valid CRS URI (or interestingly WKT) is swapped in.

After looking at this, I think I must conclude that this isn't a bug in pyproj/rasterio/rio-tiler. It is totally possible to define a tilematrixset or COG that uses a CRS that may use a geographic CRS that won't resolve with `to_authority` despite only minor modifications. 

Lowering the confidence parameter used in rio-tiler could help in some cases but won't work universally and could cause some other unintended issues. So it made the most sense to open an issue here and discuss solutions.

I think then there are 2/3 options

1) raise an error when `CRS_to_uri` evals to `None`  
2) raise a warning in the logs when this occurs and continue to allow `None` to pass through
3) Fall back to the WKT 

After reviewing OGC 06-121r9 (sections 10.2-10.3) it does look like for the most part clients would be expecting this CRS parameter to be a URI/URN. However, while testing I was able to use the full WKT string for the crs parameter and QGIS/GDAL was able to use it correctly. Given the ubiquity of GDAL, my expectation would be that this could be seen as a sensible fall back even if it's technically out-of-spec with the OGC, although maybe this is just plainly a bad idea... 

My suspicion then is that option 1 would also be too aggressive of a fix and that it might just be best to keep the change in behavior minimal by logging a warning. 

I am happy to implement this, but wanted to check with others before starting to see if there are other ideas or ways to work with the WMTS spec in ways I hadn't considered

**Comment by AndrewAnnex** [comment]
oh oops, maybe this should be labeled as an enhancement issue...

**Comment by vincentsarago** [comment]
> After reviewing OGC 06-121r9 (sections 10.2-10.3) it does look like for the most part clients would be expecting this CRS parameter to be a URI/URN. However, while testing I was able to use the full WKT string for the crs parameter and QGIS/GDAL was able to use it correctly. Given the ubiquity of GDAL, my expectation would be that this could be seen as a sensible fall back even if it's technically out-of-spec with the OGC, although maybe this is just plainly a bad idea...

Considering WMTS is a >10 yr old spec and not really up to date I would say: Let's do what ever we think is good for the users :-) 

I'm fine having WMTS response that are not 💯 following the specs especially if it's for `specific` uses case like yours where you know what you are doing ;-)

**Comment by vincentsarago** [comment]
FYI: in the new OGC Tiles `/tiles` and `/tiles/{tileMatrixSetId}` endpoints we use `CRS_to_uri` 

https://github.com/developmentseed/titiler/blob/22e48f0c17784de5c32900b789cc4acd1aab9fb3/src/titiler/core/titiler/core/factory.py#L568

for the bounding box. The OGC tiles specs allows the CRS to be either a URI or a wkt https://schemas.opengis.net/tms/2.0/json/crs.json

**Comment by AndrewAnnex** [comment]
okay excellent, I will work on this and double check things with qgis (rasterio can also probably use the wmts.xml files also to make sure everything is working).

**Comment by AndrewAnnex** [comment]
by the way @vincentsarago, have you reconsidered #546 ? The reason for me using wmts is to quickly demonstrate mosaic.json capabilities to users more familiar with QGIS. Now that QGIS has support via the OGCAPI gdal driver for OGC tiles it would make it even easier.

**Comment by vincentsarago** [comment]
> by the way @vincentsarago, have you reconsidered https://github.com/developmentseed/titiler/issues/546

What reconsideration? I've just added `tilesets` endpoint and we already had `tilematrixset` endpoints. I'm not sure what we can do more

**Comment by AndrewAnnex** [comment]
hmm okay I thought the last comment on that issue was indicating a won't fix. 

But in my testing I also wasn't able to use QGIS or the GDAL OGCAPI successfully, either way I got an error relating to a missing "bbox" parameter. Looking at GDAL's source there is a lookup for that key in the "coverage" section of the driver, but I didn't investigate further and I haven't really read into the coverages aspect of ogcapi yet. In QGIS I did select to use `TILES` in the api field also with no success.

Do you have a working example with `gdalinfo OGCAPI:https://titiler.xyz/...` handy?

**Comment by AndrewAnnex** [comment]
okay well I may have misinterpreted that comment regarding what the OGC Tiles endpoints are for then. In anycase to solve the question for this issue something like `((CRS_to_uri(crs) or crs.to_wkt()) or WGS84_CRS)` could work

**Comment by vincentsarago** [comment]
@AndrewAnnex do we have anything to do here?

**Comment by AndrewAnnex** [comment]
okay yes there would need to be a PR to implement the options I suggested in the first posting or the most recent idea for `((CRS_to_uri(crs) or crs.to_wkt()) or WGS84_CRS)`, I was waiting for #1052 to get merged before implementing, but I am about to go on vacation for a week and a bit and probably won't get to this until the new year.

**Comment by assetsrecovery** [comment]
SCAMS CAN TAKE MANY FORMS, BUT ALL SHARE A COMMON GOAL: TAKING YOUR MONEY

Fraud can happen in many ways—whether it's binary options scams, love scams, forex fraud, crypto investment schemes, or loan scams. If you’ve been a victim of any of these, it's easy to feel confused and overwhelmed by the different financial institutions and legal processes involved in recovery. The journey to reclaim your stolen funds can often be frustrating and difficult to manage on your own. 

The recovery process usually involves navigating complex legal and financial systems, which can be even more challenging without experience. In such cases, it’s crucial to seek professional help from a trusted firm specializing in fund recovery. 

“ **YOUR CHANCES OF RECOVERY DEPENDS ON THE RIGHT FIRM** “

The success of recovering your stolen funds largely depends on the firm you choose to work with. Unfortunately, the internet is full of fraudulent companies pretending to be recovery firms. These scammers often use misleading tactics and false promises to deceive victims, claiming they can recover lost funds when in fact, they are just after more of your money.

“ **BE CAUTIOUS OF IMPERSONATORS AND SCAMS** ”

Many scammers impersonate legitimate firms and use deceptive tactics such as creating fake email addresses with numbers or claiming to have recovered funds for others. It's essential to be vigilant and carefully select a reputable firm with proven expertise.

“ **WARNING: Five Websites to Avoid** ”

Here are five of the most disputed websites that you should avoid at all costs:

1. **ROYALCFD.COM**

2. **APP.DHY-GLOBAL.US**

3. **SCAMRECTIFY.ORG**

4. **INVESTCRYPTO.ORG**

5. **CRYPTOEXCHANGE.COM**

These websites are often linked to fraudulent activities and should not be trusted. Always verify the legitimacy of any platform before engaging with them.

**Official Contact Information**

Please be advised that we only operate through one official website and email address:

- Website: **[assetsretrievalboard.com](http://assetsretrievalboard.com)**
- Email: **info@assetsretrievalboard.com**

Our team includes experienced professionals who will thoroughly investigate your case and utilize advanced artificial intelligence strategies to help resolve your issue and recover your funds.

**Rest Assured, Your Recovery is Our Priority**

If you’ve fallen victim to fraud, don’t try to handle the recovery process on your own. Trust an experienced and reputable firm to maximize your chances of success.

© 2025. [Assetsretrievalboard.com]
 All rights reserved

**Comment by wakame1367** [comment]
Hi @AndrewAnnex 👋 I opened a PR implementing the WKT fallback discussed here
(option 3, with a warning): https://github.com/developmentseed/titiler/pull/1448

Happy to adjust if you were still planning to work on this or had a different
direction in mind.
